File: android/app/src/main/java/com/flutcloud/flutlink/core/SessionManager.kt
package com.flutcloud.flutlink.core

import com.flutcloud.flutlink.data.AuthSession
import com.flutcloud.flutlink.data.FlutCloudApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Holds the active account + session for the UI. Kept out of the repository
 * layer so network code only ever sees an immutable [AuthSession].
 */
class SessionManager(private val accountStore: AccountStore) {

    private val _session = MutableStateFlow<AuthSession?>(null)
    val session: StateFlow<AuthSession?> = _session.asStateFlow()

    private val _accounts = MutableStateFlow<List<AccountMeta>>(emptyList())
    val accounts: StateFlow<List<AccountMeta>> = _accounts.asStateFlow()

    suspend fun init(ocsApi: FlutCloudApi) {
        val accounts = accountStore.loadAccounts()
        _accounts.value = accounts
        restoreSession()
        refreshAdminFlags(ocsApi)
    }

    /**
     * Re-evaluate the admin flag of every stored account against the server.
     * Mirrors the desktop client's `refresh_admin_flags`: the stored flag is
     * only overwritten when the OCS probe succeeds, so a transient network
     * error never demotes an admin account to a regular one. Persists and
     * emits only when something actually changed.
     */
    suspend fun refreshAdminFlags(ocsApi: FlutCloudApi) {
        val updated = _accounts.value.map { account ->
            val token = accountStore.tokenFor(account) ?: return@map account
            val session = AuthSession(account.instanceUrl, account.username, token)
            runCatching { ocsApi.isAdmin(session) }
                .fold(
                    onSuccess = { isAdmin -> account.copy(isAdmin = isAdmin) },
                    onFailure = { account }
                )
        }
        if (updated != _accounts.value) {
            updateAccounts(updated)
        }
    }

    private suspend fun restoreSession() {
        val active = _accounts.value.firstOrNull { it.isActive } ?: _accounts.value.firstOrNull()
        if (active == null) {
            _session.value = null
            return
        }
        val token = accountStore.tokenFor(active)
        _session.value = token?.let { AuthSession(active.instanceUrl, active.username, it) }
    }

    fun addAccount(meta: AccountMeta, token: String) {
        val withoutOld = _accounts.value.filterNot {
            it.username == meta.username && it.instanceUrl.trimEnd('/') == meta.instanceUrl.trimEnd('/')
        }
        val updated = (withoutOld + meta.copy(isActive = true))
            .map { it.copy(isActive = it.username == meta.username) }
        accountStore.saveAccounts(updated)
        accountStore.saveToken(meta, token)
        _accounts.value = updated
        _session.value = AuthSession(meta.instanceUrl, meta.username, token)
    }

    fun updateAccounts(accounts: List<AccountMeta>) {
        accountStore.saveAccounts(accounts)
        _accounts.value = accounts
    }

    suspend fun switchAccount(meta: AccountMeta) {
        val updated = _accounts.value.map { it.copy(isActive = it.username == meta.username) }
        updateAccounts(updated)
        restoreSession()
    }

    fun signOut() {
        _session.value = null
    }

    fun removeAccount(meta: AccountMeta) {
        accountStore.deleteToken(meta)
        val updated = _accounts.value.filterNot {
            it.username == meta.username && it.instanceUrl.trimEnd('/') == meta.instanceUrl.trimEnd('/')
        }
        updateAccounts(updated)
        if (updated.isEmpty()) {
            _session.value = null
        } else {
            if (_session.value?.username == meta.username) {
                val next = updated.firstOrNull { it.isActive } ?: updated.first()
                val token = accountStore.tokenFor(next)
                _session.value = token?.let { AuthSession(next.instanceUrl, next.username, it) }
            }
        }
    }
}

File: ios/FlutLink/Core/SessionManager.swift
// TEST — iOS port created by opencode. Not a production build.
import Foundation
import Combine

final class SessionManager: ObservableObject {
    private let accountStore: AccountStore

    @Published private(set) var session: AuthSession?
    @Published private(set) var accounts: [AccountMeta] = []

    init(accountStore: AccountStore) {
        self.accountStore = accountStore
    }

    @MainActor
    func init_() async {
        accounts = accountStore.loadAccounts()
        restoreSession()
        await refreshAdminFlags()
    }

    @MainActor
    func refreshAdminFlags() async {
        var updated = accounts
        for (index, account) in accounts.enumerated() {
            guard let token = accountStore.tokenFor(account) else { continue }
            let s = AuthSession(baseUrl: account.instanceUrl, username: account.username, token: token)
            let api = FlutCloudApi(session: s)
            let isAdmin = await api.isAdmin()
            updated[index] = AccountMeta(
                username: account.username,
                instanceUrl: account.instanceUrl,
                displayName: account.displayName,
                isAdmin: isAdmin,
                isActive: account.isActive
            )
        }
        if updated != accounts {
            updateAccounts(updated)
        }
    }

    @MainActor
    func addAccount(meta: AccountMeta, token: String) {
        var withoutOld = accounts.filter {
            !($0.username == meta.username && $0.instanceUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) == meta.instanceUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
        }
        let updatedMeta = AccountMeta(
            username: meta.username,
            instanceUrl: meta.instanceUrl,
            displayName: meta.displayName,
            isAdmin: meta.isAdmin,
            isActive: true
        )
        withoutOld.append(updatedMeta)
        let updated = withoutOld.map { AccountMeta(username: $0.username, instanceUrl: $0.instanceUrl, displayName: $0.displayName, isAdmin: $0.isAdmin, isActive: $0.username == meta.username) }
        accountStore.saveAccounts(updated)
        accountStore.saveToken(meta, token: token)
        accounts = updated
        session = AuthSession(baseUrl: meta.instanceUrl, username: meta.username, token: token)
    }

    @MainActor
    func updateAccounts(_ accounts: [AccountMeta]) {
        accountStore.saveAccounts(accounts)
        self.accounts = accounts
    }

    @MainActor
    func switchAccount(_ meta: AccountMeta) {
        let updated = accounts.map { AccountMeta(username: $0.username, instanceUrl: $0.instanceUrl, displayName: $0.displayName, isAdmin: $0.isAdmin, isActive: $0.username == meta.username) }
        updateAccounts(updated)
        restoreSession()
    }

    @MainActor
    func signOut() {
        let updated = accounts.map { AccountMeta(username: $0.username, instanceUrl: $0.instanceUrl, displayName: $0.displayName, isAdmin: $0.isAdmin, isActive: false) }
        updateAccounts(updated)
        session = nil
    }

    @MainActor
    func removeAccount(_ meta: AccountMeta) {
        accountStore.deleteToken(meta)
        let updated = accounts.filter {
            !($0.username == meta.username && $0.instanceUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) == meta.instanceUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
        }
        updateAccounts(updated)
        if updated.isEmpty {
            session = nil
        } else if session?.username == meta.username {
            let next = updated.first(where: \.isActive) ?? updated[0]
            if let token = accountStore.tokenFor(next) {
                session = AuthSession(baseUrl: next.instanceUrl, username: next.username, token: token)
            }
        }
    }

    private func restoreSession() {
        guard let active = accounts.first(where: \.isActive),
              let token = accountStore.tokenFor(active) else {
            session = nil
            return
        }
        session = AuthSession(baseUrl: active.instanceUrl, username: active.username, token: token)
    }
}
